### PR TITLE
CMake: Implementing optional deploy of demo data

### PR DESCRIPTION
**Pull request by winterheart**
Now CMake project supports `-DENABLE_DEMO` option which controls fetching and installing HoMM II demo data.

### CODE REVIEW

**Review by oleg-derevenetz — COMMENTED** [review]
Although demo data deployment works perfectly by itself, there is one issue - fheroes2 binary is unable to find deployed files after `cmake --install` phase. That's because, although the CMake variables `CONFIGURE_FHEROES2_LOCALEDIR` and `CONFIGURE_FHEROES2_DATA` were set in CMakeLists.txt, they were never turned into corresponding preprocessor definitions. Proposed commit fixes this (along with adding proper quoting around their string values - so as not to leave it at the mercy of shell, there are many different shells and quoting rules may change from one shell to another).

P.S. There is one issue in my approach too - it assumes that `CONFIGURE_FHEROES2_*` variables contain relative paths, so `${CMAKE_INSTALL_PREFIX}` is prepended to them. This may break if they will contain absolute paths instead. Need to think how to correctly convert CMake variables to preprocessor definitions considering both possibility of relative path and possibility of absolute path. Any ideas?

File: CMakeLists.txt [hunk 1]
```diff
@@ -17,6 +17,7 @@ option(ENABLE_UNICODE  "Enable Unicode support (requires SDL/SDL2 TTF and iconv
 option(ENABLE_AUDIOCD  "Enable AudioCD support" OFF)
 option(ENABLE_TOOLS    "Enable additional tools" OFF)
 option(ENABLE_RELEASE  "Release mode (disables debug options)" ON)
+option(ENABLE_DEMO     "Fetch and install HoMM II demo data" OFF)
```
Comment (ihhub) [inline CMakeLists.txt:20]: Let's call it please `GET_HOMM2_DEMO`. People might get confused by this option without reading the description.

File: src/Makefile.switch [hunk 2]
```diff
@@ -1,8 +1,8 @@
 include $(DEVKITPRO)/libnx/switch_rules
 
 ARCH = -march=armv8-a+crc+crypto -mtune=cortex-a57 -mtp=soft -fPIE
-CFLAGS := $(CFLAGS) $(ARCH) -D__SWITCH__ -DCONFIGURE_FHEROES2_DATA=\"//switch/fheroes2\"
-CXXFLAGS := $(CXXFLAGS) $(ARCH) -D__SWITCH__ -DCONFIGURE_FHEROES2_DATA=\"//switch/fheroes2\"
+CFLAGS := $(CFLAGS) $(ARCH) -D__SWITCH__ -DCONFIGURE_FHEROES2_DATA=//switch/fheroes2
+CXXFLAGS := $(CXXFLAGS) $(ARCH) -D__SWITCH__ -DCONFIGURE_FHEROES2_DATA=//switch/fheroes2
```
Comment (ihhub) [inline src/Makefile.switch:5]: @dimag0g , could you please verify this change?

File: src/fheroes2/system/settings.cpp [hunk 3]
```diff
@@ -33,6 +33,9 @@
 #include "tinyconfig.h"
 #include "version.h"
 
+#define STRINGIFY( DEF ) #DEF
+#define EXPANDDEF( DEF ) STRINGIFY( DEF )
```
Comment (ihhub) [inline src/fheroes2/system/settings.cpp:37]: Why do we need 2 macros if we can use one single `EXPANDDEF` calls `STRINGIFY`?

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @winterheart , I left few comments here. Would you mind to take a look?

File: src/fheroes2/system/settings.cpp (hunk 3 above)
Comment (oleg-derevenetz) [reply src/fheroes2/system/settings.cpp:37]: @ihhub AFAIK when you use a macro as an argument for another macro (like here) it's only substituted when it is not used with # or ##, so we need additional macro EXPANDDEF that will expand initial macro first and then call STRINGIFY that will add quotes. May be I'm wrong of course, but you can check it :)

Comment (ihhub) [reply src/fheroes2/system/settings.cpp:37]: You're totally right. It works only with ##... I need more rest :)

**Comment by ihhub** [comment]
@winterheart , thank you very much for this improvement!

**Pull request merged**